test.rb

```diff
@@ -0,0 +1,406 @@
+              arraypush(h)
+              h= { :student_id => 2, :department => "a1", :maths => 66, :physics => 52 , :chemistry => 65, :year => 2016 }
+              arraypush(h)
+              h= { :student_id => 3, :department => "a7", :maths => 87, :physics => 32 , :chemistry => 43, :year => 2016 }
+              arraypush(h)
+              h= { :student_id => 1, :department => "a1", :maths => 21, :physics => 52 , :chemistry => 65, :year => 2015 }
+              arraypush(h)
+              h= { :student_id => 2, :department => "a1", :maths => 68, :physics => 50 , :chemistry => 65, :year => 2015 }
+              arraypush(h)
+              h= { :student_id => 3, :department => "a7", :maths => 85, :physics => 22 , :chemistry => 43, :year => 2015 }
+              arraypush(h)
+              h= { :student_id => 4, :department => "a7", :maths => 21, :physics => 22 , :chemistry => 13, :year => 2016 }
+              arraypush(h)
+    end
+
+    def arraypush(hashobject)
+      @hash_array.push(hashobject)
+    end
+
+    # Group_by method
+       def group_hashes(group_by)
+           # Group by Student_ID
+             @hash_array_groupby=Array.new
+         if(group_by.downcase=="student_id")
+
+            htemp_groupby=Hash.new
+            @hash_array= @hash_array.sort_by { |hsh| hsh[:student_id] }
+            math_total= @hash_array[0][:maths]
+            physics_total=@hash_array[0][:physics]
+            chemistry_total=@hash_array[0][:chemistry]
+            count=0
+            for num in 1...@hash_array.length
+
+                    if  num< @hash_array.length && @hash_array[num][:student_id] == @hash_array[num-1][:student_id]
+
+                    math_total=math_total+ @hash_array[num][:maths]
+                    physics_total=physics_total+ @hash_array[num][:physics]
+                    chemistry_total=chemistry_total+ @hash_array[num][:chemistry]
+
+                  else
+                    # puts "#{num}  #{count} "
+                    math_average=(math_total)/(num-count)
+                    physics_average=(physics_total)/(num-count)
+                    chemistry_average=(chemistry_total)/(num-count)
+                    # puts "Phase #{num} - #{math_total} #{physics_total} #{chemistry_total} "
+                    htemp_groupby={ :student_id => @hash_array[num-1][:student_id], :maths => math_average, :physics => physics_average , :chemistry => chemistry_average }
+                    @hash_array_groupby.push(htemp_groupby)
+
+                    math_total= @hash_array[num][:maths]
+                    physics_total=@hash_array[num][:physics]
+                    chemistry_total=@hash_array[num][:chemistry]
+                    count=num
+
+                     # puts "Phase #{num} - #{math_average} #{physics_average} #{chemistry_average} "
+                  end
+
+            end
+            # puts "#{@hash_array.length}  #{count} "
+            math_average=(math_total)/(@hash_array.length-count)
```

Comment:
store (@hash_array.length-count) in a variable instead of making calculations multiple times